python/ray/data/_internal/plan.py

```diff
@@ -72,6 +72,12 @@ def __init__(
         self._snapshot_operator: Optional[LogicalOperator] = None
         self._snapshot_stats = None
         self._snapshot_bundle = None
+        # Snapshot of only metadata corresponding to the final operator's
+        # output bundles, used as the source of truth for the Dataset's schema
+        # and count. This is calculated and cached when the plan is executed as an
+        # iterator (`execute_to_iterator()`), and avoids caching
+        # all of the output blocks in memory like in `self.snapshot_bundle`.
+        self._snapshot_metadata: Optional[BlockMetadata] = None
```

Comment:
I don't have any ideas off the top of my head, but I think it'd be good if we simplify the how we cache bundles and metadata at some point. Might be confusing how `execute_to_iterator` uses `_snapshot_metadata` but `execute` doesn't.

Reply:
good point, added a TODO comment